### PR TITLE
[FIX] tf nan gradients in issue #160

### PR DESCRIPTION
**Pull request by farrell236**

### CODE REVIEW

File: geomstats/special_euclidean_group.py [hunk 1]
```diff
@@ -39,12 +39,14 @@ class SpecialEuclideanGroup(LieGroup):
...
     """
 
-    def __init__(self, n, point_type=None):
+    def __init__(self, n, point_type=None, epsilon=0.):
         assert isinstance(n, int) and n > 1
 
         self.n = n
         self.dimension = int((n * (n - 1)) / 2 + n)
 
+        self.e = epsilon
```
Comment (ninamiolane) [inline geomstats/special_euclidean_group.py]: NIT: Can we call this `self.epsilon`? thanks.

File: geomstats/special_orthogonal_group.py [hunk 2]
```diff
@@ -41,13 +41,15 @@ class SpecialOrthogonalGroup(LieGroup, EmbeddedManifold):
...
 
-    def __init__(self, n, point_type=None):
+    def __init__(self, n, point_type=None, epsilon=0.):
 
         assert isinstance(n, int) and n > 1
 
         self.n = n
         self.dimension = int((n * (n - 1)) / 2)
 
+        self.e = epsilon
```
Comment (ninamiolane) [inline geomstats/special_orthogonal_group.py]: Ditto

File: geomstats/special_orthogonal_group.py [hunk 3]
```diff
@@ -122,20 +124,18 @@ def regularize(self, point, point_type=None):
                 mask_not_0 = ~mask_0
                 mask_pi = gs.isclose(angle, gs.pi)
 
-                mask_0_float = gs.cast(mask_0, gs.float32)
-                mask_not_0_float = gs.cast(mask_not_0, gs.float32)
-                mask_pi_float = gs.cast(mask_pi, gs.float32)
+                # This avoids division by 0.
+                mask_0_float = gs.cast(mask_0, gs.float32) + self.e
+                mask_not_0_float = gs.cast(mask_not_0, gs.float32) + self.e
+                mask_pi_float = gs.cast(mask_pi, gs.float32) + self.e
```
Comment (ninamiolane) [inline geomstats/special_orthogonal_group.py]: Nice!

File: geomstats/special_orthogonal_group.py [hunk 4]
```diff
@@ -621,7 +621,7 @@ def matrix_from_rotation_vector(self, rot_vec):
             term_2 = gs.einsum('n,njk->njk', coef_2, squared_skew_rot_vec)
 
             rot_mat = term_1 + term_2
-            rot_mat = self.projection(rot_mat)
+            # rot_mat = self.projection(rot_mat)
```
Comment (ninamiolane) [inline geomstats/special_orthogonal_group.py]: Why did you comment this? It wouldn't work?

File: geomstats/backend/tensorflow.py [hunk 5]
```diff
@@ -271,3 +271,16 @@ def diagonal(x):
 
 def mean(x, axis=None):
     return tf.reduce_mean(x, axis)
+
+def sqrtm(sym_mat):
```
Comment (ninamiolane) [inline geomstats/backend/tensorflow.py]: Got it, thanks for adding this.

File: geomstats/special_orthogonal_group.py (hunk 4 above)
Comment (farrell236) [reply geomstats/special_orthogonal_group.py]: self.projection(rot_mat) doesn't seem to change the value of rot_mat, but having it causes nan gradients. This I find weird because the gradient should be 1 if input = output.

Can this be removed? Is it necessary to have it here? I'll have another look at the function

Comment (ninamiolane) [reply geomstats/special_orthogonal_group.py]: Yes the goal of this function is to project the matrix on the space of rotations: with the taylor expansions computations, we might be slightly off some times. But it makes sense that most of the time it should give out the same matrix.

If you can have a look to prevent the nan gradients, this would be awesome. Otherwise, we can remove it for now. Thanks!

Comment (farrell236) [reply geomstats/special_orthogonal_group.py]: Uncommenting this line seem to have a weird knock on effect.

d/drot_mat(self.projection(rot_mat)) here does indeed have valid gradients, however later on in the stack trace, the nan gradient comes from ```gs.linalg.svd()```. The calling function is ```rot_mat_1 = rotations.projection(rot_mat_1)``` from ```compose()``` in special_euclidean_group.py

stack trace:
screen shot 2018-11-30 at 19 53 24

at this particular point, the variables are:
```
In[19]: mat
Out[19]: 
<tf.Tensor: id=1811, shape=(10, 3, 3), dtype=float32, numpy=
array([[[ 0.8530194 ,  0.28077707, -0.4400553 ],
        [-0.05527453,  0.8867445 ,  0.45892635],
        [ 0.5189983 , -0.3673079 ,  0.7718899 ]],
       [[ 0.60996693,  0.7835997 , -0.1179544 ],
        [-0.59181696,  0.54935443,  0.5900033 ],
        [ 0.5270548 , -0.29024464,  0.7987676 ]],
       [[ 0.9687451 ,  0.247809  ,  0.0102749 ],
        [-0.23205498,  0.8902004 ,  0.39211732],
        [ 0.08795158, -0.38236097,  0.91990256]],
       [[ 0.76246375,  0.6171269 , -0.19449371],
        [-0.54446614,  0.7742758 ,  0.32280737],
        [ 0.34969562, -0.14040326,  0.9262966 ]],
       [[ 0.47509658,  0.81414807, -0.3339426 ],
        [-0.6908534 ,  0.5801279 ,  0.43167168],
        [ 0.5450771 ,  0.02542206,  0.8379781 ]],
       [[ 0.3434578 ,  0.8581064 , -0.38180718],
        [-0.5718319 ,  0.5134961 ,  0.6398979 ],
        [ 0.7450758 , -0.00164557,  0.66695905]],
       [[ 0.681041  ,  0.71583694, -0.15420283],
        [-0.29082325,  0.45748612,  0.8403483 ],
        [ 0.67206544, -0.52761   ,  0.5196929 ]],
       [[ 0.5213585 ,  0.58598685, -0.62044525],
        [-0.19700886,  0.78995097,  0.5806845 ],
        [ 0.8303147 , -0.18069135,  0.5272102 ]],
       [[ 0.5145482 ,  0.5888327 , -0.6234311 ],
        [ 0.03063468,  0.7138044 ,  0.69964874],
        [ 0.85692817, -0.37925795,  0.34917533]],
       [[ 0.8179457 ,  0.48092622, -0.3158195 ],
        [-0.01653705,  0.568082  ,  0.8227894 ],
        [ 0.5750889 , -0.6679036 ,  0.4725862 ]]], dtype=float32)>
In[20]: 
with tf.GradientTape(persistent=True) as g:
    g.watch(mat)
    mat_unitary_u, diag_s, mat_unitary_v = gs.linalg.svd(mat)
    
In[21]: g.gradient(mat_unitary_u,mat)
Out[21]: 
<tf.Tensor: id=3072, shape=(10, 3, 3), dtype=float32, numpy=
array([[[  688366.25 ,  -181486.16 ,  -555063.8  ],
        [ -122212.03 ,  -398167.3  ,   540088.06 ],
        [ -385379.53 ,   692222.44 ,  -295073.8  ]],
       [[ -436638.72 ,   101904.85 ,   264784.44 ],
        [  128466.18 ,   422660.7  ,  -517653.88 ],
        [  441158.1  ,  -401554.03 ,    22418.377]],
       [[ -932653.4  ,  -314429.6  ,  1379220.8  ],
        [  179673.34 ,  1288920.6  , -1569616.1  ],
        [ 1753460.9  , -1568214.1  ,  -300926.5  ]],
       [[  149393.6  ,   197730.52 ,  -329915.9  ],
        [  308273.88 ,  -403909.9  ,   105521.336],
        [ -305868.03 ,     9011.174,   274908.2  ]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]],
       [[  486961.   ,    83674.31 ,  -553694.4  ],
        [  285368.44 ,  -603253.56 ,   445658.47 ],
        [ -490007.7  ,   257031.75 ,   154184.16 ]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]],
       [[  838334.7  ,   -16162.967, -1089337.2  ],
        [  104436.48 ,  -698609.   ,   626184.06 ],
        [ -699519.8  ,  1007322.06 ,  -177956.94 ]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]]], dtype=float32)>
In[22]: g.gradient(mat_unitary_v,mat)
Out[22]: 
<tf.Tensor: id=3162, shape=(10, 3, 3), dtype=float32, numpy=
array([[[-105116.05  ,   24896.812 ,   86037.28  ],
        [  18356.195 ,   54821.28  ,  -82538.21  ],
        [  63923.61  , -105823.37  ,   50229.773 ]],
       [[ 139834.19  ,  -38203.637 ,  -83675.31  ],
        [ -41355.45  , -138835.4   ,  161888.39  ],
        [-133920.58  ,  126864.45  ,   -1930.7699]],
       [[-177521.42  ,  -56761.01  ,  257528.16  ],
        [  34609.07  ,  246761.05  , -294873.94  ],
        [ 327149.56  , -295892.22  ,  -59276.43  ]],
       [[ -66518.71  ,  -98331.41  ,  158507.52  ],
        [-149679.83  ,  185701.27  ,  -55390.637 ],
        [ 152550.14  ,   -6515.452 , -128932.98  ]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]],
       [[-181772.22  ,  -25993.451 ,  212547.56  ],
        [-110789.35  ,  229837.1   , -176610.52  ],
        [ 189590.9   ,  -97743.2   ,  -65810.11  ]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]],
       [[-286907.2   ,    2058.065 ,  372538.2   ],
        [ -34684.19  ,  232112.62  , -215736.52  ],
        [ 246829.81  , -344810.4   ,   65008.824 ]],
       [[         nan,          nan,          nan],
        [         nan,          nan,          nan],
        [         nan,          nan,          nan]]], dtype=float32)>
In[23]: g.gradient(diag_s,mat)
Out[23]: 
<tf.Tensor: id=3253, shape=(10, 3, 3), dtype=float32, numpy=
array([[[ 0.8529854 ,  0.28079855, -0.43996376],
        [-0.05518094,  0.8867406 ,  0.45896217],
        [ 0.51900953, -0.3672102 ,  0.7718715 ]],
       [[ 0.60997856,  0.7836043 , -0.11785534],
        [-0.5917179 ,  0.5493447 ,  0.5899918 ],
        [ 0.5270635 , -0.29014516,  0.79876137]],
       [[ 0.96873796,  0.24786979,  0.01035395],
        [-0.23196024,  0.8901792 ,  0.39214176],
        [ 0.08798333, -0.38228428,  0.9198465 ]],
       [[ 0.7624599 ,  0.6171434 , -0.194395  ],
        [-0.5443732 ,  0.7742398 ,  0.32281715],
        [ 0.34973288, -0.14031151,  0.9262828 ]],
       [[        nan,         nan,         nan],
        [        nan,         nan,         nan],
        [        nan,         nan,         nan]],
       [[ 0.3434787 ,  0.85809064, -0.3817107 ],
        [-0.5717342 ,  0.51349616,  0.6398768 ],
        [ 0.74507916, -0.0015468 ,  0.6669745 ]],
       [[        nan,         nan,         nan],
        [        nan,         nan,         nan],
        [        nan,         nan,         nan]],
       [[        nan,         nan,         nan],
        [        nan,         nan,         nan],
        [        nan,         nan,         nan]],
       [[ 0.51454353,  0.58881295, -0.6233332 ],
        [ 0.03073353,  0.7138182 ,  0.6996564 ],
        [ 0.8569133 , -0.37916094,  0.34919417]],
       [[        nan,         nan,         nan],
        [        nan,         nan,         nan],
        [        nan,         nan,         nan]]], dtype=float32)>
```

Is ```mat``` here caused by this [SVD gradient is unstable for non-unique singular values #17476](https://github.com/tensorflow/tensorflow/issues/17476)?

Sample 5,7,8,10 return nan from svd. What are your thoughts on this?

Comment (ninamiolane) [reply geomstats/special_orthogonal_group.py]: Thank you so much for the debug! It's incredibly useful.

If it comes from the svd, it's slightly more worrying as we use this function several times (here, and in stiefel.py) :/  

As I understand it, the lack of gradient comes from the fact that the svd function is multi-valued, in terms of the unit eigen vectors, when mat is: some constant * identity matrix. Thus, the lack of gradient is a mathematical restriction.

However, it seems that some people found a work-around for deep learning applications:
https://github.com/tensorflow/tensorflow/issues/6503

Is there a way we could force the autograd to use this?

Comment (farrell236) [reply geomstats/special_orthogonal_group.py]: I had a look through that issue, it seems to be quite old. I can't seem to find the PR page for it (i.e. if it ever made into tensorflow master branch). However, the solution provided doesn't solve the issue in our case, it is still a TODO :/ ([Link](https://github.com/caisq/tensorflow/commit/e1d9e4ed05420ab9486bf18ec331e90e59e51982#diff-8bc1f62af8f7af2dd411f1845ac5ecedR305) for the commit with the feature fix). In file ```tensorflow/python/ops/linalg_grad.py``` there is the segment:

```
   # TODO(rmlarsen): Define a gradient that is numerically stable for
     # abs(m-n) > 1. Currently this does not work because there are effectively
     # multiple singular values with value zero. I am not sure if this is a true
     # instability or if it simply throws off the finite difference gradient
     # checker.
     if abs(m - n) > 1:
       raise NotImplementedError(
           "svd gradient is not implemented for abs(m - n) > 1")
     s_mat = array_ops.matrix_diag(s)
     s2 = math_ops.square(s)
```

In tensorflow 1.10.1 (the current one I'm using) have this in its place:
```
    # NOTICE: Because of the term involving f, the gradient becomes
    # infinite (or NaN in practice) when singular values are not unique.
    # Mathematically this should not be surprising, since for (k-fold)
    # degenerate singular values, the corresponding singular vectors are
    # only defined up a (k-dimensional) subspace. In practice, this can
    # lead to numerical instability when singular values are close but not
    # exactly equal.
```

Comment (farrell236) [reply geomstats/special_orthogonal_group.py]: JaeDukSeo was quoting kcyu2014 from above, unfortunately in that implementation, there's no gradient for U matrix. 

I agree, we should discuss the SVD issue in a separate issue thread/PR. This seems it may be a big task on its own. Is it ok to comment out or disable other instances of self.projection() until a sound solution is found?

Comment (ninamiolane) [reply geomstats/special_orthogonal_group.py]: Yes, let's remove this (not commenting) for now. Thanks!
